server/src/test/java/org/elasticsearch/search/crossproject/ProjectTagsTests.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.search.crossproject;
+
+import org.elasticsearch.cluster.metadata.ProjectId;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ProjectTagsTests extends ESTestCase {
+
+    public void testValidTags() {
+        ProjectId projectId = randomUniqueProjectId();
+        ProjectTags.validateTags(projectId.id(), randomTags(projectId));
+    }
+
+    public void testMissingTags() {
+        ProjectId projectId = randomUniqueProjectId();
+        Map<String, String> tags = randomTags(projectId);
+        tags.remove(randomFrom("_id", "_alias", "_type", "_organization"));
```

Comment:
could this be exhaustive rather than randomised?

Reply:
for sure. we used to abuse randomization for this sort of thing -- old habits die hard.